### PR TITLE
[Example/WASM] Support four types of quantized models

### PR DESCRIPTION
**Pull request by Wenzhao-Xiang**

### CODE REVIEW

**Comment by Wenzhao-Xiang** [comment]
1. support quantized Mobilenet V1/V2 and SSD Mobilenet V1/V2. And the model resources are in `model/README.md`.
2. support  Uint8 version of Conv2D/DepthwiseConv2D/Averagepool/Reshape/Softmax on WASM backend.

__Test Summary for WASM__

Regression: 0

before:
passes: 469 failures: 246

after:
passes: 472 failures: 243

**Comment by Wenzhao-Xiang** [comment]
Please go https://github.com/intel/webml-polyfill/issues/650#issuecomment-475308591 for demo URL.

**Comment by Wenzhao-Xiang** [comment]
@huningxin @ibelem Please help take a look. Thanks!

File: src/nn/wasm/PreparedModel.js [hunk 1]
```diff
@@ -111,24 +113,148 @@ export default class PreparedModel {
...
-        throw new Error("Unsupported fused activation function.");
+        throw new Error("Unsupported type of tensor for fused activation function.");
+      }
+    }
+
+    function GetQuantizedConvolutionMultipler(input_scale, filter_scale, 
+                                              bias_scale, output_scale) {
+      let input_product_scale = input_scale * filter_scale;
+      // const float bias_scale = biasShape.scale;
+      // const float output_scale = outputShape.scale;
```
Comment (huningxin) [inline src/nn/wasm/PreparedModel.js]: Please remove the commented code.

File: src/nn/wasm/PreparedModel.js [hunk 2]
```diff
@@ -111,24 +113,148 @@ export default class PreparedModel {
...
+      right_shift *= -1;
+      quantized_multiplier = Math.round(q * -(1 << 31));
+      OPS_CHECK(quantized_multiplier <= -(1 << 31));
+      if (quantized_multiplier == -(1 << 31)) {
+        quantized_multiplier /= 2;
+        --right_shift;
+      }
+      OPS_CHECK(right_shift >= 0);
+      OPS_CHECK(quantized_multiplier <= nn_ops.INT32_MAX);
+      // quantized_multiplier = static_cast<int32_t>(q_fixed);
```
Comment (huningxin) [inline src/nn/wasm/PreparedModel.js]: ditto.

File: src/nn/wasm/PreparedModel.js [hunk 3]
```diff
@@ -111,24 +113,148 @@ export default class PreparedModel {
...
+        ++left_shift;
+      }
+      OPS_CHECK(left_shift >= 0);
+      OPS_CHECK(quantized_multiplier <= nn_ops.INT32_MAX);
+      return [quantized_multiplier, left_shift];
+    }
+
+    function CalculateInputRadius(input_integer_bits, input_left_shift) {
+      let max_input_rescaled = 1.0 * ((1 << input_integer_bits) - 1) *
+                                        (1 << (31 - input_integer_bits)) /
```
Comment (huningxin) [inline src/nn/wasm/PreparedModel.js]: keep indent to above line.

File: src/nn/wasm/PreparedModel.js [hunk 4]
```diff
@@ -514,17 +747,37 @@ export default class PreparedModel {
...
         let output = operands[outputs[0]];
 
+        let inputMultiplier = 0, inputLeftShift = 0, diffMin = 0;
+        if (output.type === OperandCode.TENSOR_QUANT8_ASYMM) {
+          if (output.zeroPoint != 0 || output.scale != 1 / 256) {
+            console.error("incorrect scale / offset for output");
+          }
+          let kScaledDiffIntegerBits = 5;
+          let input_beta_real_multiplier =
+                  Math.min(1.0 * beta * input.scale * (1 << (31 - kScaledDiffIntegerBits)), -(1 << 31) - 1.0);
```
Comment (huningxin) [inline src/nn/wasm/PreparedModel.js]: keep 4-spaces indent to above line.

File: src/nn/wasm/PreparedModel.js [hunk 5]
```diff
@@ -111,24 +113,148 @@ export default class PreparedModel {
...
+      let max_input_rescaled = 1.0 * ((1 << input_integer_bits) - 1) *
+                                        (1 << (31 - input_integer_bits)) /
+                                        (1 << input_left_shift);
+      // Tighten bound using floor.  Suppose that we could use the exact value.
+      // After scaling the difference, the result would be at the maximum.  Thus we
+      // must ensure that our value has lower magnitude.
+      return Math.floor(max_input_rescaled);
+    }
+
+    function frexp (arg) {
```
Comment (huningxin) [inline src/nn/wasm/PreparedModel.js]: Is the code converted from C counterpart? If yes, please add reference to C function. Thanks!

**Comment by huningxin** [comment]
> 2\. support  Uint8 version of Conv2D/DepthwiseConv2D/Averagepool/Reshape/Softmax on WASM backend.

Looks good with some nits. Please take a look.



> 1. support quantized Mobilenet V1/V2 and SSD Mobilenet V1/V2. And the model resources are in `model/README.md`.

@ibelem , please take a look. Thanks!

**Comment by ibelem** [comment]
I know the conflict will happen when #662 was merged, @Wenzhao-Xiang please fix the conflict.

**Comment by Wenzhao-Xiang** [comment]
Fixed conflicts and optimized the code style. @huningxin @ibelem PLAT, thanks!

**Comment by ibelem** [comment]
Merge this PR, thanks @Wenzhao-Xiang for the great work!

**Pull request merged**